relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -0,0 +1,441 @@
+                SamplingRule {
+                    project_ids: vec![project_id],
+                    releases: GlobPatterns::new(vec!["1.1.1".to_string()]),
+                    user_segments: vec!["vip".to_string()],
+                    sample_rate: 1.0,
+                },
+            ),
+            (
+                "multiple projects",
+                SamplingRule {
+                    project_ids: vec![project_id2, project_id, project_id3],
+                    releases: GlobPatterns::new(vec!["1.1.1".to_string()]),
+                    user_segments: vec!["vip".to_string()],
+                    sample_rate: 1.0,
+                },
+            ),
+            (
+                "all projects",
+                SamplingRule {
+                    project_ids: vec![],
+                    releases: GlobPatterns::new(vec!["1.1.1".to_string()]),
+                    user_segments: vec!["vip".to_string()],
+                    sample_rate: 1.0,
+                },
+            ),
+            (
+                "glob releases",
+                SamplingRule {
+                    project_ids: vec![project_id],
+                    releases: GlobPatterns::new(vec!["1.*".to_string()]),
+                    user_segments: vec!["vip".to_string()],
+                    sample_rate: 1.0,
+                },
+            ),
+            (
+                "multiple releases",
+                SamplingRule {
+                    project_ids: vec![project_id],
+                    releases: GlobPatterns::new(vec!["2.1.1".to_string(), "1.1.*".to_string()]),
+                    user_segments: vec!["vip".to_string()],
+                    sample_rate: 1.0,
+                },
+            ),
+            (
+                "multiple user segments",
+                SamplingRule {
+                    project_ids: vec![project_id],
+                    releases: GlobPatterns::new(vec!["1.1.1".to_string()]),
+                    user_segments: vec!["paid".to_string(), "vip".to_string(), "free".to_string()],
+                    sample_rate: 1.0,
+                },
+            ),
+            (
+                "all user segments",
+                SamplingRule {
+                    project_ids: vec![project_id],
+                    releases: GlobPatterns::new(vec!["1.1.1".to_string()]),
+                    user_segments: vec![],
+                    sample_rate: 1.0,
+                },
+            ),
+            (
+                "match all",
+                SamplingRule {
+                    project_ids: vec![],
+                    releases: GlobPatterns::new(vec![]),
+                    user_segments: vec![],
+                    sample_rate: 1.0,
+                },
+            ),
+        ];
+
+        for (rule_test_name, rule) in rules.iter() {
+            let failure_name = format!("Failed on test: '{}'!!!", rule_test_name);
+            assert!(matches(rule, &trace_context, project_id), failure_name);
```

Comment:
_nit_: You an also generate test functions with a macro, in which case they are executed independently of each other.

Reply:
Yes doing a loop in a test is a poor substitution for proper data fixtures in the unit test framework.
I think unit test should be as plane as possible, I don't want to use macros unless they are part of a well documented test framework.
I would rather go for using some already made test library/framework that has **D**ata **D**riven **T**est features.

Reply:
Works for me, thanks for clarifying.

To close on this, here is an example of how this looks with macros (except that the macro would include the function):
https://github.com/getsentry/relay/blob/b352da85ce00f8280662f9ca30aab637c7a56dfe/relay-common%2Fsrc%2Fglob.rs#L92-L126